starknet_py/proxy_check.py

```diff
@@ -16,9 +13,11 @@ async def is_proxy(self, contract: "Contract") -> bool:
         """
 
     @abstractmethod
-    async def implementation_address(self, contract: "Contract") -> int:
+    async def implementation(self, contract: "Contract") -> int:
         """
-        :return: Address of contract being proxied by `contract` given as a parameter
+        :return: Implementation (either class hash or contract address)
+            of contract being proxied by `contract` given as a parameter
```

Comment:
Allowing this to be either class hash or contract address complicates the handling of the proxies. I think we could extend the class hierarchy a little bit and add something along the lines of `ProxyCheckAddress` and `ProxyCheckClassHash` as the subclasses of the `ProxyCheck` and then handle them accordingly based on their type (with `isinstance` checks).

Reply:
I think we should resolve the abi for an address, why would we need to resolve for class hash?